<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Общая-информация" data-toc-modified-id="Общая-информация-1">Общая информация</a></span><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1.1">Обзор данных</a></span></li></ul></li><li><span><a href="#Анализ" data-toc-modified-id="Анализ-2">Анализ</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-выпущенных-после-после-01/01/2020" data-toc-modified-id="Количество-книг,-выпущенных-после-после-01/01/2020-2.1">Количество книг, выпущенных после после 01/01/2020</a></span></li><li><span><a href="#Количество--обзоров-и-средняя-оценка-на-книгу" data-toc-modified-id="Количество--обзоров-и-средняя-оценка-на-книгу-2.2">Количество  обзоров и средняя оценка на книгу</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3">Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг(-расчет-по-книгам-с-50-и-более-оценками)" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг(-расчет-по-книгам-с-50-и-более-оценками)-2.4">Автор с самой высокой средней оценкой книг( расчет по книгам с 50 и более оценками)</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-2.5">Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3">Выводы</a></span></li></ul></div>

# Анализ базы данных сервиса для чтения книг.

В нашем распоряжении база данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг  

**Цель исследования:**
 * поиск инсайтов для формирования предложения для нового продукта, связанного с чтением книг по подписке

## Общая информация

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
pd.set_option('max_colwidth', 120)

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

**Функция**

In [3]:
def sql_query(query):
    display(pd.io.sql.read_sql(query, con = engine))

### Обзор данных

**Таблица `books`**

In [4]:
books = '''
SELECT *
FROM books
LIMIT 5
''' 

In [5]:
sql_query(books)

book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268



- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

In [6]:
authors = '''
SELECT *
FROM authors
LIMIT 5
''' 

In [7]:
sql_query(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd



- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

In [8]:
publishers = '''
SELECT *
FROM publishers
LIMIT 5
''' 

In [9]:
sql_query(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company




- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

In [53]:
ratings = '''
SELECT *
FROM ratings
LIMIT 5
''' 

In [11]:
sql_query(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2



- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

In [12]:
reviews = '''
SELECT *
FROM reviews
LIMIT 5
''' 

In [13]:
sql_query(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                      text  
0                               Mention society tell send professor analysis. Over provide race technology continue these.  
1                                        Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social par...  
3  Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play n...  
4         Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.



- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Анализ

### Количество книг, выпущенных после после 01/01/2020

In [14]:
count_books = '''
SELECT COUNT(book_id) as cnt
FROM books
WHERE publication_date > '2000-01-01'
''' 

In [15]:
sql_query(count_books)

cnt
0  819

**819 книг вышло после 1 января 2000 года**

### Количество  обзоров и средняя оценка на книгу

In [51]:
reviews_books = '''
WITH rw AS(
        SELECT book_id,
        COUNT(review_id) AS cnt
        FROM reviews
        GROUP BY book_id
            ),
        rt AS(
        SELECT book_id,
        ROUND(AVG(rating),2) AS avg_rating
        FROM ratings
        GROUP BY book_id
            )            
            
SELECT b.book_id,
        b.title,
        COALESCE(rw.cnt,0) AS cnt,
        rt.avg_rating
FROM books AS b
LEFT JOIN rw AS rw ON b.book_id = rw.book_id
LEFT JOIN rt AS rt ON b.book_id = rt.book_id
ORDER BY cnt DESC, rt.avg_rating DESC
LIMIT 15
''' 

In [52]:
sql_query(reviews_books)

book_id                                                        title  cnt  \
0       948                                      Twilight (Twilight  #1)    7   
1       302  Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)    6   
2       299   Harry Potter and the Chamber of Secrets (Harry Potter  #2)    6   
3       656                                               The Book Thief    6   
4       734                                             The Glass Castle    6   
5       497                                    Outlander (Outlander  #1)    6   
6       750                          The Hobbit  or There and Back Again    6   
7       779    The Lightning Thief (Percy Jackson and the Olympians  #1)    6   
8       695            The Curious Incident of the Dog in the Night-Time    6   
9       963                                          Water for Elephants    6   
10      696                       The Da Vinci Code (Robert Langdon  #2)    6   
11      673                                       The Catcher in the Rye    6   
12      627                                                The Alchemist    6   
13      854                                                     The Road    6   
14      733                                    The Giver (The Giver  #1)    6   

    avg_rating  
0         3.66  
1         4.41  
2         4.29  
3         4.26  
4         4.21  
5         4.13  
6         4.13  
7         4.08  
8         4.08  
9         3.98  
10        3.83  
11        3.83  
12        3.79  
13        3.77  
14        3.75

**Больше всего обзоров пишут на фэнтэзи, в том числе с аудиторией для подростков - Сумерки, серия про Гарри Поттера, Хоббит Туда и Обратно.  
Самые высокие рейтинги из этих книг у серии про Гарри Поттера**

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [18]:
publishers = '''
SELECT p.publisher_id,
        p.publisher,
        COUNT(b.book_id) AS cnt
FROM publishers AS p 
JOIN books AS b ON p.publisher_id=b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id,p.publisher
ORDER BY cnt DESC
LIMIT 1
''' 

In [19]:
sql_query(publishers)

publisher_id      publisher  cnt
0           212  Penguin Books   42

**Издательство `Penguin Books` выпустило наибольшее число книг толще 50 страниц (исключили брошюры)**

### Автор с самой высокой средней оценкой книг( расчет по книгам с 50 и более оценками)

In [30]:
top_author = '''
WITH b50 AS(
            SELECT b.book_id
            FROM books AS b 
            JOIN ratings AS r ON b.book_id=r.book_id
            GROUP BY b.book_id
            HAVING COUNT(rating) > 50
            )

SELECT a.author_id,
    a.author,
    ROUND(AVG(rating),2) AS avg_rating
FROM books AS b 
JOIN ratings AS r ON b.book_id=r.book_id
JOIN authors AS a ON b.author_id=a.author_id
WHERE b.book_id IN (SELECT book_id FROM  b50)
GROUP BY a.author_id, a.author 
ORDER BY avg_rating DESC
LIMIT 1

'''

In [31]:
sql_query(top_author)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré        4.29

**Автор книг с самыми высокими рейтингами Джоан Роулинг (Мэри Гранпрэ - иллюстратор серии про Гарри Поттера)**

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [56]:
reviewers = '''
WITH rev50 AS (
                SELECT username,
                        COUNT(rating_id) AS cnt
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) >50
                ),
    rev AS (
                SELECT username,
                        COUNT(review_id) AS cnt
                FROM reviews
                GROUP BY username
                )
SELECT ROUND(AVG(rev.cnt),0) AS review_avg
FROM rev50
JOIN rev ON rev50.username=rev.username
'''

In [57]:
sql_query(reviewers)

review_avg
0        24.0

**В среднем 24 обзора написали пользователи, которые поставили книгам больше 50 оценок**

## Выводы

 * 819 книг вышло после 1 января 2000 года
 * Больше всего обзоров написано на жанр фэнтэзи, в том числе с аудиторией для подростков 
 * Издательство Penguin Books выпустило наибольшее число книг (без учета брошюр)
 * Джоан Роулинг - автор книг с самыми высокими рейтингами  
 * 24 обзора в среднем написали пользователи, которые поставили книгам больше 50 оценок
  